### PR TITLE
Updated ConcatOptimization to support Concat with 0D input of one dimension

### PR DESCRIPTION
**Pull request by lazarevevgeny**
Description: Updated ConcatOptimization to support Concat with 0D input of one dimension

JIRA: 37875

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR
* [x]  Transformation preserves original framework node names

Validation:
* [x]  Unit tests: Done
* [x]  Framework operation tests: N/A no new op enabled
* [x]  Transformation tests: Done
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py: N/A no new public model
* [x]  Model Optimizer IR Reader check: Done

Documentation:
* [x]  Supported frameworks operations list: N/A no new public model
* [x]  Supported **public** models list: : N/A no new public model
* [x]  New operations specification: N/A no new public model
* [x]  Guide on how to convert the **public** model: N/A no new public model
* [x]  User guide update: N/A

### CODE REVIEW

**Review by jane-intel — APPROVED** [review]
Are we supposed to re-connect Concat input ports when deleted one from the middle? Just asking

**Comment by lazarevevgeny** [comment]
> Are we supposed to re-connect Concat input ports when deleted one from the middle? Just asking

Yes we are. And I think that this is correct behaviour. At least I don't see a reason why this will be a problem (the only scenario in my head is when the tensor with 0 dimension is marked as output for some reason).

**Comment by jane-intel** [comment]
> > Are we supposed to re-connect Concat input ports when deleted one from the middle? Just asking
> 
> Yes we are. And I think that this is correct behaviour. At least I don't see a reason why this will be a problem (the only scenario in my head is when the tensor with 0 dimension is marked as output for some reason).

Then I don't see us doing it. Is it port renumbering? or we really reconnect inputs manually?

**Comment by lazarevevgeny** [comment]
> > > Are we supposed to re-connect Concat input ports when deleted one from the middle? Just asking
> > 
> > 
> > Yes we are. And I think that this is correct behaviour. At least I don't see a reason why this will be a problem (the only scenario in my head is when the tensor with 0 dimension is marked as output for some reason).
> 
> Then I don't see us doing it. Is it port renumbering? or we really reconnect inputs manually?

My bad. I initially implemented the transformation with ports reconnection at the back stage but then found this transformation in the middle and updated it but didn't pay attention to the fact that it does not reconnect inputs. Now code performs inputs re-connection.

File: model-optimizer/extensions/middle/ConcatOptimization.py [hunk 1]
```diff
@@ -107,13 +103,18 @@ class ConcatOdInputEraser(MiddleReplacementPattern):
...
-                shape = in_port.data.get_shape()
-                assert shape is not None
-                if np.array_equal(shape, [0]):
-                    in_port.disconnect()
-
-            connected_input_ports = [in_port for in_port in concat.in_ports().values() if not in_port.disconnected()]
-            assert len(connected_input_ports), 'Concat {} does nothing'.format(concat.soft_get('name', concat.id))
+            port_to_connect = 0
+            for port_idx in range(len(concat.in_ports())):
+                if concat.is_in_port_connected(port_idx) and 0 in concat.in_port(port_idx).data.get_shape():
```
Comment (lazarevevgeny) [inline model-optimizer/extensions/middle/ConcatOptimization.py:108]: Bug if the `port_idx` is not connected.

File: model-optimizer/extensions/middle/ConcatOptimization.py [hunk 2]
```diff
@@ -112,8 +108,19 @@ def find_and_replace_pattern(self, graph: Graph):
                     continue
                 shape = in_port.data.get_shape()
                 assert shape is not None
-                if np.array_equal(shape, [0]):
+                if 0 in shape:
                     in_port.disconnect()
 
             connected_input_ports = [in_port for in_port in concat.in_ports().values() if not in_port.disconnected()]
             assert len(connected_input_ports), 'Concat {} does nothing'.format(concat.soft_get('name', concat.id))
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/ConcatOptimization.py:115]: I believe this message is confusing. Could you please correct it?

File: model-optimizer/extensions/middle/ConcatOptimization.py [hunk 3]
```diff
@@ -112,8 +108,19 @@ def find_and_replace_pattern(self, graph: Graph):
...
+
+            max_port_id = max([port.idx for port in connected_input_ports])
+            if max_port_id != len(connected_input_ports) + 1:
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/ConcatOptimization.py:118]: Is this `if True:` statement? 

Lets say my_concat have connected in ports with indices `[0, 1, 2, 3]`, than `max_port_id == 3`  and `(len(connected_input_ports) + 1)==5`. Algorithm will try to reconnect ports, but everything is fine there.

Maybe we could be more straight forward like this:
`[port.idx for port in connected_input_ports] != list(range(len(connected_input_ports)))`

**Review by jane-intel — CHANGES_REQUESTED** [review]

File: model-optimizer/extensions/middle/ConcatOptimization.py (hunk 2 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/middle/ConcatOptimization.py]: Done

File: model-optimizer/extensions/middle/ConcatOptimization.py (hunk 3 above)
Comment (lazarevevgeny) [reply model-optimizer/extensions/middle/ConcatOptimization.py]: Updated. Actually we need to check that if number of connected ports has changed or their numbers have changed then we need to re-order inputs. So I have updated the check.

File: model-optimizer/extensions/middle/ConcatOptimization.py [hunk 4]
```diff
@@ -98,22 +94,39 @@ def find_and_replace_pattern(self, graph: Graph):
...
-            connected_input_ports = [in_port for in_port in concat.in_ports().values() if not in_port.disconnected()]
-            assert len(connected_input_ports), 'Concat {} does nothing'.format(concat.soft_get('name', concat.id))
+            connected_ports = [port for port_idx, port in sorted(concat.in_ports().items()) if not port.disconnected()]
+            assert len(connected_ports), 'Concat "{}" have no inputs after removing inputs with 0 dimensions' \
+                                         ''.format(concat.soft_get('name', concat.id))
+
+            max_port_id = max([port.idx for port in connected_ports])
+            # if we renumbered input ports or removed some of them we need to re-connect the inputs
+            if [port.idx for port in connected_ports] != list(range(len(connected_ports))) or \
+                    num_connected_ports != len(connected_ports):
```
Comment (jane-intel) [inline model-optimizer/extensions/middle/ConcatOptimization.py:124]: Why do we need the second check? First one covers all cases as I understand.

Comment (lazarevevgeny) [reply model-optimizer/extensions/middle/ConcatOptimization.py:124]: No, it does not. Refer to test case `test_deletion_3`

**Review by jane-intel — APPROVED** [review]

**Pull request merged**